### PR TITLE
Add support for the Ornata Chroma V2

### LINKED ISSUE #1187: Support for Ornata V2
### Device Info

* Name: Ornata V2
* Product Number: RZ03-03380100-R3M1
* Model Number: RZ03-0338

### Device Information
```
Bus 001 Device 009: ID 1532:025d Razer USA, Ltd 
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x025d 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Ornata V2
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x006d
    bNumInterfaces          4
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     179
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              2 Razer Ornata V2
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval               1
Device Status:     0x0000
  (Bus Powered)

```
```
❯     ls -lah /dev/input/by-id/

total 0
drwxr-xr-x 2 root root 260 jul  1 20:46 .
drwxr-xr-x 4 root root 700 jul  1 20:46 ..
lrwxrwxrwx 1 root root   9 jun 29 19:17 usb-Azurewave_Integrated_Camera_0001-event-if00 -> ../event5
lrwxrwxrwx 1 root root  10 jul  1 20:46 usb-Razer_Razer_Ornata_V2-event-if01 -> ../event23
lrwxrwxrwx 1 root root  10 jul  1 20:46 usb-Razer_Razer_Ornata_V2-event-if03 -> ../event25
lrwxrwxrwx 1 root root  10 jul  1 20:46 usb-Razer_Razer_Ornata_V2-event-kbd -> ../event10
lrwxrwxrwx 1 root root  10 jul  1 20:46 usb-Razer_Razer_Ornata_V2-if01-event-kbd -> ../event11
lrwxrwxrwx 1 root root  10 jul  1 20:46 usb-Razer_Razer_Ornata_V2-if02-event-mouse -> ../event24
lrwxrwxrwx 1 root root   9 jul  1 20:46 usb-Razer_Razer_Ornata_V2-if02-mouse -> ../mouse3
lrwxrwxrwx 1 root root   9 jun 29 19:17 usb-Razer_Razer_Ouroboros-event-mouse -> ../event4
lrwxrwxrwx 1 root root   9 jun 29 19:17 usb-Razer_Razer_Ouroboros-if01-event-kbd -> ../event8
lrwxrwxrwx 1 root root   9 jun 29 19:17 usb-Razer_Razer_Ouroboros-if02-event-kbd -> ../event9
lrwxrwxrwx 1 root root   9 jun 29 19:17 usb-Razer_Razer_Ouroboros-mouse -> ../mouse0
```
### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

not able to capture packets atm

### PR DESCRIPTION
**Pull request by ShimmerGlass**
Add support for the Ornata Chroma V2, based on existing keyboards.

What works :
* Effect "none"
* Effect "spectrum"
* Effect "wave"
* Effect "reactive"
* Effect "breath"
* Effect "starlight"
* Effect "static"
* Custom frame write and display
* Luminosity
* Game mode

What doesn't : 
* Effect "ripple" (doesn't seem implemented for other keyboards)

Fixes: #1187

### CODE REVIEW

**Comment by Hellrazor1215** [comment]
I am very new at this....I download the files...not sure what to do with it
thank you for your great work!

**Comment by swedneck** [comment]
Has anyone else been able to get this to work?

**Comment by Hellrazor1215** [comment]
@swedneck I have been trying for the last five days. With no luck so far.

**Comment by TermoZour** [comment]
> @swedneck I have been trying for the last five days. With no luck so far.

What exactly have you been trying?

**Comment by Hellrazor1215** [comment]
@TermoZour I am trying to install the new driver for my keyboard

**Comment by TermoZour** [comment]
> @TermoZour I am trying to install the new driver for my keyboard

Yes but what were the exact steps you've been trying? That's what I meant. All we know is you downloaded some files.

**Comment by Anty0** [comment]
Can confirm driver does work. Tested with RazerGenie, razerCommander and polychromatic.

Tested:
- Effect "none"
- Effect "spectrum"
- Effect "wave"
- Effect "reactive"
- Effect "breath"
- Effect "starlight"
- Effect "static"
- Luminosity

Doesn't work:
- Wave with direction to the left (only works direction to the right)

**Comment by nightsky30** [comment]
> * Wave with direction to the left (only works direction to the right)

@Anty0 It has been found that with the chroma extended matrix wave effect some devices use the wave left/right direction values `(0x00, 0x01)` while other use `(0x01, 0x02)`. If you tweak the direction values to `WAVE_DIRS = (1, 2)`, rebuild, etc, do both directions start working for you?

https://github.com/openrazer/openrazer/blob/9440bfee527d505010582bca8448a9bc87ad0e69/daemon/openrazer_daemon/hardware/keyboards.py#L618-L620

**Comment by Hellrazor1215** [comment]
Of course TermoZour, a read the Makefile and install (or I think I did). make daemon_install, make driver_install, make udev_install and make setup_dkms. Clean install as I try to install to a new linux MX-Linux installation. But that did not work. So I try installing the package from the repo and then use the command make driver_install. I did clean between my tries. Hope it clear enough, it is the first time I use git, please be patient.

**Comment by nightsky30** [comment]
@Hellrazor1215 I'm not familiar with MX Linux, but it seems to be Mepis/Debian based, as are Ubuntu/Mint. That being said, you could try the build instructions [here](https://github.com/openrazer/openrazer/wiki/Building-a-package). If the Ubuntu/Mint instructions don't work, you could also try the Debian instructions at the bottom of that page (very similar).

**Comment by swedneck** [comment]
i'm on fedora 32 and i haven't been able to get it to work, building and installing the rpm just results in the keyboard not working at all until i remove the driver.

**Comment by sergeLabo** [comment]
Work's on Debian 10

git clone -b ornata-chroma-v2 https://github.com/ShimmerGlass/openrazer.git
cd openrazer
sed -i '/linux-headers-generic/d' debian/control
./scripts/build_debs.sh
sudo dpkg -i ./dist/*.deb

Restart

**Comment by Hellrazor1215** [comment]
Thanks guys works great. followed sergeLabo instructions and notice that I was missing debhelper and python3-setuptools package. After installing these packages everything was fine and after a restart my keyboard was detected. Thanks again, now I have to find how to setup my own custom scheme.

File: README.md [hunk 1]
```diff
@@ -35,6 +35,7 @@ The devices below are fully feature supported by OpenRazer, which means all avai
 | Razer BlackWidow X Ultimate                   |  1532:0217  |
 | Razer BlackWidow X Tournament Edition Chroma  |  1532:021A  |
 | Razer Ornata Chroma                           |  1532:021E  |
+| Razer Ornata Chroma V2                        |  1532:025D  |
```
Comment (z3ntu) [inline README.md]: Please sort by VID (put before Razer Cynosa V2)

File: driver/razerkbd_driver.h [hunk 2]
```diff
@@ -37,6 +37,7 @@
 #define USB_DEVICE_ID_RAZER_BLACKWIDOW_X_ULTIMATE 0x0217
 #define USB_DEVICE_ID_RAZER_BLACKWIDOW_X_CHROMA_TE 0x021A
 #define USB_DEVICE_ID_RAZER_ORNATA_CHROMA 0x021E
+#define USB_DEVICE_ID_RAZER_ORNATA_CHROMA_V2 0x025D
```
Comment (z3ntu) [inline driver/razerkbd_driver.h]: sort by vid (put before USB_DEVICE_ID_RAZER_CYNOSA_V2)

File: install_files/udev/99-razer.rules [hunk 3]
```diff
@@ -10,7 +10,7 @@ ATTRS{idProduct}=="0013|0016|0020|0024|0025|002e|002f|0032|0034|0036|0037|0038|0
     ENV{ID_RAZER_CHROMA}="1", ENV{RAZER_DRIVER}="razermouse"
 
 # Keyboards
-ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|0252|0253|0255|025e", \
+ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|0252|0253|0255|025e|025d", \
```
Comment (z3ntu) [inline install_files/udev/99-razer.rules]: put 025e before 025d (sort)

**Review by z3ntu — CHANGES_REQUESTED** [review]
some minor things, LGTM otherwise

**Comment by swedneck** [comment]
any ideas how to troubleshoot why this version's driver just makes my keyboard not send any input at all?

**Comment by Hellrazor1215** [comment]
> i'm on fedora 32 and i haven't been able to get it to work, building and installing the rpm just results in the keyboard not working at all until i remove the driver.

swedneck make sure there is no error when you compile, had the same problem on debian, some packages were needed to compile correctly.

**Comment by Anty0** [comment]
> > * Wave with direction to the left (only works direction to the right)
> 
> @Anty0 It has been found that with the chroma extended matrix wave effect some devices use the wave left/right direction values `(0x00, 0x01)` while other use `(0x01, 0x02)`. If you tweak the direction values to `WAVE_DIRS = (1, 2)`, rebuild, etc, do both directions start working for you?
> 
> https://github.com/openrazer/openrazer/blob/9440bfee527d505010582bca8448a9bc87ad0e69/daemon/openrazer_daemon/hardware/keyboards.py#L618-L620


Can confirm changing `WAVE_DIRS` as suggested by @nightsky30 fixes wave effect for me. Tested with polychromatic on [this](https://github.com/Anty0/openrazer/tree/ornata-chroma-v2) branch.

Just as a side note, for some reason the name of my keyboard (on the box) is Ornata V2 (without Chroma). (I'm not sure if there are more versions of this keyboard...)

**Comment by AtlasC0R3** [comment]
> Just as a side note, for some reason the name of my keyboard (on the box) is Ornata V2 (without Chroma). (I'm not sure if there are more versions of this keyboard...)


No, you aren't alone with this box. I think it's just the official box that has that name but the Synapse software on Windows has Ornata Chroma V2. _Probably because they couldn't fit it all on the box._

**Comment by swedneck** [comment]
@Hellrazor1215 i can't see any errors when building the packages with `rpmbuild`

**Comment by ShimmerGlass** [comment]
@z3ntu rebased and fixed the sorting issues
@nightsky30 changed WAVE_DIRS as suggested

**Comment by kevchu3** [comment]
> > > * Wave with direction to the left (only works direction to the right)
> > 
> > 
> > @Anty0 It has been found that with the chroma extended matrix wave effect some devices use the wave left/right direction values `(0x00, 0x01)` while other use `(0x01, 0x02)`. If you tweak the direction values to `WAVE_DIRS = (1, 2)`, rebuild, etc, do both directions start working for you?
> > https://github.com/openrazer/openrazer/blob/9440bfee527d505010582bca8448a9bc87ad0e69/daemon/openrazer_daemon/hardware/keyboards.py#L618-L620
> 
> Can confirm changing `WAVE_DIRS` as suggested by @nightsky30 fixes wave effect for me. Tested with polychromatic on [this](https://github.com/Anty0/openrazer/tree/ornata-chroma-v2) branch.
> 
> Just as a side note, for some reason the name of my keyboard (on the box) is Ornata V2 (without Chroma). (I'm not sure if there are more versions of this keyboard...)

I'm on Fedora 33 and pulled the latest changes (followed the wiki as usual) and it compiles.  Tested the waves in both directions on my Ornata v2, and they look fine to me.  Good work, @ShimmerGlass

**Review by z3ntu — APPROVED** [review]
LGTM, thanks!

**Comment by nyc-97** [comment]
I tried following sergeLabo's instructions and this doesn';t work for me. The ornata-chrome-v2 branch has been deleted. It says its supported on the website but Im not getting my keyboard recognized. my mouse and actual laptop keyboard is working fine though. Not sure whats going on.

**Pull request merged**